ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java

```diff
@@ -0,0 +1,155 @@
+ */
+
+package tech.pegasys.artemis.statetransition;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import net.consensys.cava.bytes.Bytes;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.util.BeaconStateUtil;
+import tech.pegasys.artemis.datastructures.util.ValidatorsUtil;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+
+public class LmdGhost {
+
+  public BeaconBlock lmd_ghost(
+      ChainStorageClient store, BeaconState start_state, BeaconBlock start_block)
+      throws StateTransitionException {
+    List<Integer> active_validator_indices =
+        ValidatorsUtil.get_active_validator_indices(
+            start_state.getValidator_registry(),
+            BeaconStateUtil.slot_to_epoch(UnsignedLong.valueOf(start_block.getSlot())));
+
+    List<BeaconBlock> attestation_targets = new ArrayList<>();
+    for (Integer validatorIndex : active_validator_indices) {
+      try {
+        attestation_targets.add(get_latest_attestation_target(store, validatorIndex));
+      } catch (Error err) {
+        System.err.println(err);
+      }
+    }
+
+    BeaconBlock head = start_block;
+    List<BeaconBlock> children;
+    while (true) {
+      children = get_children(store, head);
+
+      if (children.size() == 0) {
+        return head;
+      }
+
+      head =
+          children.stream()
+              .max(
+                  Comparator.comparing(
+                      child_block ->
+                          Math.toIntExact(
+                              get_vote_count(store, child_block, attestation_targets).longValue())))
+              .get();
+    }
+  }
+
+  /*
+   * This function is defined inside lmd_ghost in spec. It is defined here separately for legibility.
+   */
+  public UnsignedLong get_vote_count(
+      ChainStorageClient store, BeaconBlock block, List<BeaconBlock> attestation_targets) {
+    UnsignedLong vote_count = UnsignedLong.ZERO;
+    for (BeaconBlock target : attestation_targets) {
+      Optional<BeaconBlock> ancestor =
+          get_ancestor(store, target, UnsignedLong.valueOf(block.getSlot()));
+      if (!ancestor.isPresent()) continue;
+      if (ancestor.get().equals(block)) {
+        vote_count = vote_count.plus(UnsignedLong.ONE);
+      }
+    }
+    return vote_count;
+  }
+
+  /*
+   * Spec pseudo-code:
+   *  Let get_children(store: Store, block: BeaconBlock) -> List[BeaconBlock] returns
+   *  the child blocks of the given block.
+   */
+  // TODO: OPTIMIZE: currently goes through all the values in processedBlockLookup
+  public List<BeaconBlock> get_children(ChainStorageClient store, BeaconBlock block) {
```

Comment:
ouch...good call on the TODO.  i'm thinking that a good optimization would be a new mapping with key = block_root and value = [childblock1,childblock2,...,childblockN] would be worth it.

Reply:
I agree. That brings up the question: where do we save that? every time we receive a new block?